management/backup.py

```diff
@@ -513,79 +534,176 @@ def list_target_files(config):
-		min_age = int(min_age)
+	return "OK"
 
-	config["target"] = target
-	config["target_user"] = target_user
-	config["target_pass"] = target_pass
-	config["min_age_in_days"] = min_age
+def set_local_backup_config(env, min_age_in_days):
+	# min_age_in_days must be an int
+	if isinstance(min_age_in_days, str):
+		min_age_in_days = int(min_age_in_days)
+
+	config = {
+		"type": "local",
+		"min_age_in_days": min_age_in_days
+	}
+
+	write_backup_config(env, config)
+
+	return "OK"
+
+def set_rsync_backup_config(env, min_age_in_days, target_url):
+	# min_age_in_days must be an int
+	if isinstance(min_age_in_days, str):
+		min_age_in_days = int(min_age_in_days)
+
+	config = {
+		"type": "rsync",
+		"target_url": target_url,
+		"min_age_in_days": min_age_in_days
+	}
 
 	# Validate.
 	try:
-		if config["target"] not in ("off", "local"):
-			# these aren't supported by the following function, which expects a full url in the target key,
-			# which is what is there except when loading the config prior to saving
-			list_target_files(config)
+		list_target_files(config)
 	except ValueError as e:
 		return str(e)
 
 	write_backup_config(env, config)
 
 	return "OK"
 
-def get_backup_config(env, for_save=False, for_ui=False):
-	backup_root = os.path.join(env["STORAGE_ROOT"], 'backup')
+def set_s3_backup_config(env, min_age_in_days, s3_access_key_id, s3_secret_access_key, target_url, s3_endpoint_url, s3_region_name=None):
+	# min_age_in_days must be an int
+	if isinstance(min_age_in_days, str):
+		min_age_in_days = int(min_age_in_days)
+
+	config = {
+		"type": "s3",
+		"target_url": target_url,
+		"s3_endpoint_url": s3_endpoint_url,
+		"s3_region_name": s3_region_name,
+		"s3_access_key_id": s3_access_key_id,
+		"s3_secret_access_key": s3_secret_access_key,
+		"min_age_in_days": min_age_in_days
+	}
+
+	if s3_region_name is not None:
+		config["s3_region_name"] = s3_region_name
+
+	# Validate.
+	try:
+		list_target_files(config)
+	except ValueError as e:
+		return str(e)
+
+	write_backup_config(env, config)
+
+	return "OK"
+
+def set_b2_backup_config(env, min_age_in_days, target_url):
+	# min_age_in_days must be an int
+	if isinstance(min_age_in_days, str):
+		min_age_in_days = int(min_age_in_days)
+
+	config = {
+		"type": "b2",
+		"target_url": target_url,
+		"min_age_in_days": min_age_in_days
+	}
+
+	# Validate.
+	try:
+		list_target_files(config)
+	except ValueError as e:
+		return str(e)
+
+	write_backup_config(env, config)
+
+	return "OK"
+
+def get_backup_config(env):
+	backup_root = get_backup_root_directory(env)
 
 	# Defaults.
 	config = {
-		"min_age_in_days": 3,
-		"target": "local",
+		"type": "local",
+		"min_age_in_days": 3
 	}
 
 	# Merge in anything written to custom.yaml.
 	try:
 		with open(os.path.join(backup_root, 'custom.yaml'), 'r') as f:
 			custom_config = rtyaml.load(f)
 		if not isinstance(custom_config, dict): raise ValueError() # caught below
+
+		# Converting the previous configuration (which was not very clear)
```

Comment:
The appropriate place for this code is `setup/migrate.py`, it's cleaner if the config file is changed once at upgrade time and `backup.py` itself only knows how to interpret the current valid file format

Reply:
I tried to move this code from `management/backup.py` to `setup/migrate.py` but I bumped into the impossibility of using `rtyaml` and `urllib` modules in `setup/migrate.py` since they are loaded later. Do you have any suggestions to give me?

Reply:
@maaarghk can you give me any suggestions?

Reply:
I'm not sure but looking [here](https://github.com/mail-in-a-box/mailinabox/blob/main/setup/start.sh#L32) seems to indicate that `migrate.py --migrate` is run after requirements are installed, where as [here](https://github.com/mail-in-a-box/mailinabox/blob/main/setup/start.sh#L87) i can see only `migrate.py --current` is run before requirements are installed. So, it should work to do something like [this](https://github.com/mail-in-a-box/mailinabox/blob/main/setup/migrate.py#L156) i.e. conditionally import rtyaml / urllib only if the backup config file is found in the user data directory.

Reply:
Solved with 607ad27941a227ca54c23b1199589537e902bc02.

In the end I solved by falling back on the `yaml` module instead of `rtyaml`, since the latter is not available. [Here is the list of installed modules](https://github.com/mail-in-a-box/mailinabox/files/10702116/modules.txt) obtained with `help('modules')`.